### PR TITLE
Bert large uncased eager

### PR DESCRIPTION
**Pull request by IvanShynkarenko**
### Ticket
[GitHub Issue #1073](https://github.com/tenstorrent/pytorch2.0_ttnn/issues/1073)

### Problem Description
Enable the full forward pass of Hugging Face’s `bert-large-uncased` model to run on the `ttnn:0` device using native TT-NN mappings.

### What's Changed
1. Implemented required mapped operations.
2. Added layout compatibility logic to convert tensors from **Row Major** to **Tiled**, as expected by TTNN ops.
3. Integrated model-to-device logic to ensure all tensors and submodules correctly transfer to `ttnn:0`.
4. Added a test script to verify correctness.  
   ➤ Script path: `tests/cpp_extension/test_bert_cpp_extension_eager.py`

### CODE REVIEW

File: tools/run_transformers.py [hunk 1]
```diff
@@ -188,7 +188,7 @@ def run_model(
         TestModel("state-spaces/mamba-130m-hf", AutoModelForCausalLM, "Hey how are you doing?"),
         TestModel("huggyllama/llama-7b", AutoModelForCausalLM, "Spring is a good time to"),
         TestModel(
-            "mnoukhov/gpt2-imdb-sentiment-classifier",
+            "mnoukhov/gpt2-imdb-sentiment-  ",
```
Comment (ayerofieiev-tt) [inline tools/run_transformers.py:191]: why?

File: torch_ttnn/cpp_extension/ttnn_cpp_extension/include/ttnn_cpp_extension/ops/view.hpp [hunk 2]
```diff
@@ -0,0 +1,45 @@
...
+#include <c10/core/SymInt.h>
+
+namespace tt_eager::ops::view {
+
+using TtnnTensor = ttnn::Tensor;
+using Shape = ttnn::Shape;
+
+TtnnTensor select(const TtnnTensor& input, int dim, int index);
+
+TtnnTensor broadcast(const TtnnTensor& input, const Shape& target_shape);
```
Comment (ayerofieiev-tt) [inline torch_ttnn/cpp_extension/ttnn_cpp_extension/include/ttnn_cpp_extension/ops/view.hpp:20]: should these also have a `ttnn_` prefix?

File: torch_ttnn/cpp_extension/ttnn_cpp_extension/include/ttnn_cpp_extension/ops/view.hpp [hunk 3]
```diff
@@ -0,0 +1,45 @@
...
+
+at::Tensor ttnn_view(const at::Tensor& self, at::IntArrayRef size);
+
+at::Tensor ttnn_expand(const at::Tensor& self, at::IntArrayRef size, bool implicit);
+
+at::Tensor ttnn_permute(const at::Tensor& input, at::IntArrayRef dims);
```
Comment (ayerofieiev-tt) [inline torch_ttnn/cpp_extension/ttnn_cpp_extension/include/ttnn_cpp_extension/ops/view.hpp:26]: should different functions here go to different headers as you have it for other ops like norm, matmul, etc?

File: torch_ttnn/cpp_extension/ttnn_cpp_extension/src/open_registration_extension.cpp [hunk 4]
```diff
@@ -10,20 +10,52 @@
...
+    m.impl("bmm", &tt_eager::ops::binary::ttnn_bmm);
+    m.impl("clone", &tt_eager::ops::unary::ttnn_clone);
+    m.impl("div.Tensor", &tt_eager::ops::binary::ttnn_div_tensor);
+    m.impl("embedding", &tt_eager::ops::embedding::ttnn_embedding);
+    m.impl("expand", &tt_eager::ops::view::ttnn_expand);
+    m.impl("gelu", &tt_eager::ops::unary::ttnn_gelu);
+    m.impl("mul.Tensor", &tt_eager::ops::binary::ttnn_mul_tensor);
+    m.impl("native_layer_norm", &tt_eager::ops::norm::ttnn_native_layer_norm);
+    m.impl("permute", &tt_eager::ops::view::ttnn_permute);
+    m.impl("rsub.Scalar", &tt_eager::ops::binary::ttnn_rsub_scalar);
```
Comment (ayerofieiev-tt) [inline torch_ttnn/cpp_extension/ttnn_cpp_extension/src/open_registration_extension.cpp:46]: interestingly, `rsub` is not displayed here
https://docs.pytorch.org/docs/stable/torch.compiler_ir.html
same for `split.Tensor` and `t`

I guess docs are outdated

File: torch_ttnn/cpp_extension/ttnn_cpp_extension/src/ops/linear_iml.cpp [hunk 5]
```diff
@@ -0,0 +1,102 @@
+#include "ttnn_cpp_extension/ops/linear_iml.hpp"
```
Comment (ayerofieiev-tt) [inline torch_ttnn/cpp_extension/ttnn_cpp_extension/src/ops/linear_iml.cpp:1]: why is this file called linear_imlpl when others are going without _impl? can rename to linear.hpp/.cpp?

File: torch_ttnn/cpp_extension/ttnn_cpp_extension/src/ops/unary.cpp [hunk 6]
```diff
@@ -1,33 +1,126 @@
...
     out_tensor_impl->set_sizes_and_strides_as(self);
 
-    // Set output TTNN tensor to result
     auto out_ttnn_tensor = out_tensor_impl->get_ttnn_tensor();
     out_tensor_impl->set_ttnn_tensor(result);
 
     return out;
 }
 
+at::Tensor ttnn_clone(const at::Tensor& self, c10::optional<at::MemoryFormat> memory_format_opt) {
```
Comment (ayerofieiev-tt) [inline torch_ttnn/cpp_extension/ttnn_cpp_extension/src/ops/unary.cpp:36]: I think this belongs to its own file

usually we would put unary eltwise ops into unary.
and even though this function accepts a single tensor, it is more like a data movement operation or a core operation.

File: torch_ttnn/cpp_extension/ttnn_cpp_extension/src/ops/unary.cpp [hunk 7]
```diff
@@ -1,33 +1,126 @@
...
+    LOGGING("Running aten::clone.default");
+    TORCH_CHECK(self.device().type() == c10::DeviceType::PrivateUse1);
+
+    auto* self_impl = static_cast<at::TtnnTensorImpl*>(self.unsafeGetTensorImpl());
+    auto ttnn_tensor = self_impl->get_ttnn_tensor();
+
+    auto cloned =
+        ttnn::operations::data_movement::clone::Clone::invoke(ttnn_tensor, std::nullopt, std::nullopt, std::nullopt);
```
Comment (ayerofieiev-tt) [inline torch_ttnn/cpp_extension/ttnn_cpp_extension/src/ops/unary.cpp:44]: You are not supposed to call ::invoke function yourself. 
Just `ttnn::clone()`

see 
https://github.com/tenstorrent/tt-metal/blob/main/ttnn/cpp/ttnn/operations/data_movement/clone/clone.hpp#L19-L21

File: torch_ttnn/cpp_extension/ttnn_cpp_extension/src/ops/view.cpp [hunk 8]
```diff
@@ -0,0 +1,371 @@
...
+    auto ttnn_tensor = self_impl->get_ttnn_tensor();
+
+    if (ttnn_tensor.layout() == ttnn::ROW_MAJOR_LAYOUT) {
+        ttnn_tensor = ttnn::to_layout(ttnn_tensor, ttnn::TILE_LAYOUT, std::nullopt, std::nullopt, ttnn_tensor.device());
+    }
+
+    ttnn::SmallVector<uint32_t> target_shape_vec(size.begin(), size.end());
+    ttnn::Shape target_shape(target_shape_vec);
+
+    auto broadcasted_tensor = ttnn::experimental::broadcast_to(
```
Comment (ayerofieiev-tt) [inline torch_ttnn/cpp_extension/ttnn_cpp_extension/src/ops/view.cpp:32]: there is `ttnn::expand` which might be better to use instead I think

File: torch_ttnn/cpp_extension/ttnn_cpp_extension/src/ops/view.cpp [hunk 9]
```diff
@@ -0,0 +1,371 @@
...
+        out_sizes[i] = input.size(dims[i]);
+    }
+    auto output = tt_eager::ops::create::custom_empty_memory_format(
+        out_sizes, input.scalar_type(), c10::nullopt, input.device(), c10::nullopt);
+    auto* out_impl = static_cast<at::TtnnTensorImpl*>(output.unsafeGetTensorImpl());
+    out_impl->set_ttnn_tensor(ttnn_output);
+    return output;
+}
+
+at::Tensor ttnn_slice_tensor(
```
Comment (ayerofieiev-tt) [inline torch_ttnn/cpp_extension/ttnn_cpp_extension/src/ops/view.cpp:138]: please move into its own cpp

File: torch_ttnn/cpp_extension/ttnn_cpp_extension/src/ops/view.cpp [hunk 10]
```diff
@@ -0,0 +1,371 @@
...
+        /*pin_memory=*/c10::nullopt);
+
+    auto* out_impl = static_cast<at::TtnnTensorImpl*>(output.unsafeGetTensorImpl());
+
+    out_impl->set_ttnn_tensor(ttnn_output);
+
+    return output;
+}
+
+std::vector<at::Tensor> ttnn_split_tensor_fixed(const at::Tensor& self, c10::SymInt split_size, int64_t dim) {
```
Comment (ayerofieiev-tt) [inline torch_ttnn/cpp_extension/ttnn_cpp_extension/src/ops/view.cpp:192]: "fixed"?

File: torch_ttnn/cpp_extension/ttnn_cpp_extension/src/ops/view.cpp [hunk 11]
```diff
@@ -0,0 +1,371 @@
...
+    LOGGING("Running aten::split.Tensor (fixed size)");
+
+    TORCH_CHECK(self.device().type() == c10::DeviceType::PrivateUse1, "Tensor must be on PrivateUse1 device");
+    int64_t split_sz;
+    if (split_size.is_symbolic()) {
+        split_sz = split_size.guard_int(__FILE__, __LINE__);
```
Comment (ayerofieiev-tt) [inline torch_ttnn/cpp_extension/ttnn_cpp_extension/src/ops/view.cpp:198]: can you help to understand whats going on here?

File: torch_ttnn/cpp_extension/ttnn_cpp_extension/src/ops/view.cpp [hunk 12]
```diff
@@ -0,0 +1,371 @@
...
+    auto output = tt_eager::ops::create::custom_empty_memory_format(
+        shape_vec, self.scalar_type(), c10::nullopt, self.device(), c10::nullopt);
+
+    auto* out_impl = static_cast<at::TtnnTensorImpl*>(output.unsafeGetTensorImpl());
+    out_impl->set_ttnn_tensor(ttnn_result);
+
+    return output;
+}
+
+at::Tensor ttnn_transpose_int(const at::Tensor& input, int64_t dim0, int64_t dim1) {
```
Comment (ayerofieiev-tt) [inline torch_ttnn/cpp_extension/ttnn_cpp_extension/src/ops/view.cpp:262]: what is the practical difference between t and trasponse_int and why they can't have same impl or make one call another?

**Comment by fahmedchanTT** [comment]
@IvanShynkarenko have you made all the requested changes ? If so, please re-request a review.

File: torch_ttnn/cpp_extension/ttnn_cpp_extension/src/ops/linear_iml.cpp (hunk 5 above)
Comment (IvanShynkarenko) [reply torch_ttnn/cpp_extension/ttnn_cpp_extension/src/ops/linear_iml.cpp]: I’ve already fixed it

File: torch_ttnn/cpp_extension/ttnn_cpp_extension/src/ops/unary.cpp (hunk 7 above)
Comment (IvanShynkarenko) [reply torch_ttnn/cpp_extension/ttnn_cpp_extension/src/ops/unary.cpp]: I have changed it

File: torch_ttnn/cpp_extension/ttnn_cpp_extension/src/ops/view.cpp (hunk 8 above)
Comment (IvanShynkarenko) [reply torch_ttnn/cpp_extension/ttnn_cpp_extension/src/ops/view.cpp]: yeah, I didn’t see it;
 I’ve already changed it.

File: torch_ttnn/cpp_extension/ttnn_cpp_extension/src/ops/view.cpp [hunk 13]
```diff
@@ -0,0 +1,371 @@
...
+    auto output = tt_eager::ops::create::custom_empty_memory_format(
+        new_size, self.scalar_type(), c10::nullopt, self.device(), c10::nullopt);
+
+    auto* out_impl = static_cast<at::TtnnTensorImpl*>(output.unsafeGetTensorImpl());
+    out_impl->set_ttnn_tensor(reshaped_tensor);
+
+    return output;
+}
+
+at::Tensor ttnn_permute(const at::Tensor& input, at::IntArrayRef dims) {
```
[the remark this answers is not in the document]
Comment (IvanShynkarenko) [reply torch_ttnn/cpp_extension/ttnn_cpp_extension/src/ops/view.cpp]: done

File: torch_ttnn/cpp_extension/ttnn_cpp_extension/src/ops/view.cpp (hunk 9 above)
Comment (IvanShynkarenko) [reply torch_ttnn/cpp_extension/ttnn_cpp_extension/src/ops/view.cpp]: done

File: torch_ttnn/cpp_extension/ttnn_cpp_extension/src/open_registration_extension.cpp [hunk 14]
```diff
@@ -10,20 +10,52 @@
...
+    m.impl("slice.Tensor", &tt_eager::ops::view::ttnn_slice_tensor);
+    m.impl("split.Tensor", &tt_eager::ops::view::ttnn_split_tensor_fixed);
+    m.impl("t", &tt_eager::ops::view::ttnn_t_default);
```
[the remark this answers is not in the document]
Comment (IvanShynkarenko) [reply torch_ttnn/cpp_extension/ttnn_cpp_extension/src/open_registration_extension.cpp:55]: yes, first one work only for 2-d

File: torch_ttnn/cpp_extension/ttnn_cpp_extension/include/ttnn_cpp_extension/utils/extension_utils.hpp [hunk 15]
```diff
@@ -5,6 +5,10 @@
 #include <fmt/format.h>
 #include <ttnn/operations/core/core.hpp>
 
+namespace tt_eager {
+  using TtnnTensor = ttnn::Tensor;
+}
```
[the remark this answers is not in the document]
Comment (IvanShynkarenko) [reply torch_ttnn/cpp_extension/ttnn_cpp_extension/include/ttnn_cpp_extension/utils/extension_utils.hpp]: done

File: tools/run_transformers.py (hunk 1 above)
Comment (IvanShynkarenko) [reply tools/run_transformers.py]: sorry for that, I accidentally edited the file, changed it back

File: torch_ttnn/cpp_extension/ttnn_cpp_extension/src/ops/matmul.cpp [hunk 16]
```diff
@@ -0,0 +1,62 @@
...
+    auto ttnn_other = ensure_tile(other_impl->get_ttnn_tensor());
+
+    auto ttnn_self_shape = ttnn_self.logical_shape();
+    auto ttnn_other_shape = ttnn_other.logical_shape();
+
+    auto ttnn_result = ttnn::matmul(ttnn_self, ttnn_other);
+
+    auto ttnn_result_shape = ttnn_result.logical_shape();
+
+    std::vector<int64_t> output_shape;
```
[the remark this answers is not in the document]
Comment (IvanShynkarenko) [reply torch_ttnn/cpp_extension/ttnn_cpp_extension/src/ops/matmul.cpp]: done

File: tests/autogen_op/BERT/eager_mode/test_BERT_aten__softmax_default_eager.py [hunk 17]
```diff
@@ -0,0 +1,88 @@
+import torch
```
[the remark this answers is not in the document]
Comment (IvanShynkarenko) [reply tests/autogen_op/BERT/eager_mode/test_BERT_aten__softmax_default_eager.py:1]: yes they are autogenerated based on other tests that were already in test directory

File: torch_ttnn/cpp_extension/ttnn_cpp_extension/src/ops/view.cpp [hunk 18]
```diff
@@ -0,0 +1,371 @@
...
+    auto output = tt_eager::ops::create::custom_empty_memory_format(
+        shape_vec, self.scalar_type(), c10::nullopt, self.device(), c10::nullopt);
+
+    auto* out_impl = static_cast<at::TtnnTensorImpl*>(output.unsafeGetTensorImpl());
+    out_impl->set_ttnn_tensor(ttnn_result);
+
+    return output;
+}
+
+at::Tensor ttnn_squeeze_dim(const at::Tensor& self, int64_t dim) {
```
[the remark this answers is not in the document]
Comment (IvanShynkarenko) [reply torch_ttnn/cpp_extension/ttnn_cpp_extension/src/ops/view.cpp]: done

File: torch_ttnn/cpp_extension/ttnn_cpp_extension/include/ttnn_cpp_extension/ops/view.hpp (hunk 3 above)
Comment (IvanShynkarenko) [reply torch_ttnn/cpp_extension/ttnn_cpp_extension/include/ttnn_cpp_extension/ops/view.hpp:25]: moved functions to different headers

File: torch_ttnn/cpp_extension/ttnn_cpp_extension/src/open_registration_extension.cpp (hunk 4 above)
Comment (IvanShynkarenko) [reply torch_ttnn/cpp_extension/ttnn_cpp_extension/src/open_registration_extension.cpp:52]: probably yes

**Comment by IvanShynkarenko** [comment]
I still couldn’t manage to fine-tune the accuracy of the responses...
all other issues were closed

**Comment by marty1885** [comment]
@IvanShynkarenko is there anything still blocking and/or we can do to help getting the PR through?

**Comment by marty1885** [comment]
@IvanShynkarenko Final call. Will close and reopen next Monday
